From 77ee7e0d47a64bc734706899ac7b5442796f9420 Mon Sep 17 00:00:00 2001
From: Dinolek <git@dinolek.me>
Date: Thu, 19 Jun 2025 21:12:03 +0200
Subject: [PATCH 030/115] memory: mtk-smi: mt6785: Add SMI Support

Add MT6785 SMI larb support and init reg pair to
MT6779 SMI common, which is compatible with MT6785.

Signed-off-by: Dinolek <git@dinolek.me>
---
 drivers/memory/mtk-smi.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)
 mode change 100644 => 100755 drivers/memory/mtk-smi.c

diff --git a/drivers/memory/mtk-smi.c b/drivers/memory/mtk-smi.c
old mode 100644
new mode 100755
index aaeba8ab2..a6711b72c
--- a/drivers/memory/mtk-smi.c
+++ b/drivers/memory/mtk-smi.c
@@ -283,6 +283,22 @@ static int mtk_smi_larb_config_port_gen2_general(struct device *dev)
 	return 0;
 }
 
+static const u8 mtk_smi_larb_mt6785_ostd[][SMI_LARB_PORT_NR_MAX] = {
+	[0] = {0x2, 0x2, 0x1, 0x1f, 0x1, 0x1, 0xa, 0xa, 0x28},
+	[1] = {0x2, 0x1, 0x1c, 0x1c, 0xa, 0x1, 0x1, 0xd, 0xd, 0xe, 0x1, 0xe, 0x1,
+	       0x28},
+	[2] = {0x18, 0x1, 0x8, 0x1, 0x1, 0x1, 0x1, 0x1, 0x1, 0x2, 0x1, 0x1},
+	[3] = {0x1, 0x3, 0x2, 0x1, 0x1, 0x1, 0x1, 0x4, 0x2, 0x1, 0x4, 0x1, 0x1,
+	       0x1, 0x1, 0x4, 0xb, 0x13, 0x14},
+	[4] = {0x1, 0x1, 0x1},
+	[5] = {0xc, 0x1, 0x4, 0x4, 0x1, 0x2, 0x2, 0x6, 0x1, 0x9, 0x2, 0x2, 0x2,
+	       0x1, 0x6, 0x9, 0x1, 0x1, 0x6, 0x4, 0x4, 0x1, 0x1, 0x1, 0x1, 0x2},
+	[6] = {0x1f, 0x14, 0x2, 0x2, 0x2, 0x1, 0x16, 0x8, 0x2, 0x2, 0x2, 0x1, 0xa,
+	       0xa, 0x4, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x2, 0x1,
+	       0x1, 0x1, 0x1, 0x1, 0x1},
+	[7] = {0x1f, 0x14, 0x1f, 0x1f, 0x14},
+};
+
 static const u8 mtk_smi_larb_mt6893_ostd[][SMI_LARB_PORT_NR_MAX] = {
 	[0] = {0x2, 0x6, 0x2, 0x2, 0x2, 0x28, 0x18, 0x18, 0x1, 0x1, 0x1, 0x8,
 	       0x8, 0x1, 0x3f},
@@ -498,6 +514,12 @@ static const struct mtk_smi_larb_gen mtk_smi_larb_mt6779 = {
 		/* DUMMY | IPU0 | IPU1 | CCU | MDLA */
 };
 
+static const struct mtk_smi_larb_gen mtk_smi_larb_mt6785 = {
+	.config_port                = mtk_smi_larb_config_port_gen2_general,
+	.flags_general	            = MTK_SMI_FLAG_THRT_UPDATE | MTK_SMI_FLAG_SW_FLAG,
+	.ostd		            = mtk_smi_larb_mt6785_ostd,
+};
+
 static const struct mtk_smi_larb_gen mtk_smi_larb_mt6893 = {
 	.config_port                = mtk_smi_larb_config_port_gen2_general,
 	.flags_general	            = MTK_SMI_FLAG_THRT_UPDATE | MTK_SMI_FLAG_SW_FLAG,
@@ -549,6 +571,7 @@ static const struct of_device_id mtk_smi_larb_of_ids[] = {
 	{.compatible = "mediatek,mt2701-smi-larb", .data = &mtk_smi_larb_mt2701},
 	{.compatible = "mediatek,mt2712-smi-larb", .data = &mtk_smi_larb_mt2712},
 	{.compatible = "mediatek,mt6779-smi-larb", .data = &mtk_smi_larb_mt6779},
+	{.compatible = "mediatek,mt6785-smi-larb", .data = &mtk_smi_larb_mt6785},
 	{.compatible = "mediatek,mt6795-smi-larb", .data = &mtk_smi_larb_mt8173},
 	{.compatible = "mediatek,mt6893-smi-larb", .data = &mtk_smi_larb_mt6893},
 	{.compatible = "mediatek,mt8167-smi-larb", .data = &mtk_smi_larb_mt8167},
@@ -768,6 +791,7 @@ static const struct mtk_smi_common_plat mtk_smi_common_mt6779 = {
 	.has_gals = true,
 	.bus_sel  = F_MMU1_LARB(1) | F_MMU1_LARB(2) | F_MMU1_LARB(4) |
 		    F_MMU1_LARB(5) | F_MMU1_LARB(6) | F_MMU1_LARB(7),
+	.init     = mtk_smi_common_mt8195_init,
 };
 
 static const struct mtk_smi_common_plat mtk_smi_common_mt6795 = {
-- 
2.56.0

